okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -493,6 +493,14 @@ private void onGoAway(int lastKnownStreamId, Status status) {
     boolean notifyShutdown;
     synchronized (lock) {
       notifyShutdown = !goAway;
+    }
+    if (notifyShutdown) {
+      // We call transportShutdown() here, because once the goAway becomes true, other thread may
+      // call stopIfNecessary() and cause listener.transportTerminated() be called.
```

Comment:
The transportTerminated() is always called after the goAway becomes true, after this change, the goAway will only becomes true after transportShutdown() is called, so they should be fine.